### PR TITLE
Feature/cps sampling

### PR DESCRIPTION
**Pull request by halfflat**
Adds conditional Poisson sampling, and a new `demo_sample` demo program that examines the sampler behaviour in isolation.

Other features
* correlation reporting in `demo_distribute`
* conditional Poisson sampling implemented either as multinomial rejection or by Poisson rejection
* `demo_sample` will calculate and report second-order inclusion probabilities (useful for highlighting shortcomings in `oss` sampler)
* `counting_iterator` add to iterator utilities; gives a random-access iterator interface to an integer range.

### CODE REVIEW

File: demo/demo_distribute.cc [hunk 1]
```diff
@@ -322,6 +328,31 @@ struct running_stats {
...
+        cn+=(x-mx)*dy;
+    }
+
+    int n;
+    double mx,my,cn;
+};
+
 // harness
 
 void run_test(const cl_args &A) {
```
Comment (shurikasa) [inline demo/demo_distribute.cc]: @halfflat Could you please add some more test description? Thanks!

File: demo/demo_sample.cc [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+                << std::setprecision(4) << min_ratio << " to " << max_ratio << ".";
+    
+            throw usage_error(msg.str());
+        }
+    }
+
+    return A;
+}
+
+void run_test(const cl_args &);
```
Comment (shurikasa) [inline demo/demo_sample.cc]: @halfflat A bit more comments in this test as well please )

**Comment by halfflat** [comment]
Added some more comments: is more explanation required, do you think?

**Comment by shurikasa** [comment]
Looks very good, thank you!

**Pull request merged**